keps/262-ConfigurableFailurePolicy/README.md

```diff
@@ -0,0 +1,392 @@
+// fails due to a reason listed in OnJobFailureReasons.
+// Only only rule can match a unique (Job failure reason, ReplicatedJob) pair.
+type FailurePolicyRule struct {
+  // The action to take if the rule is matched.
+  // +kubebuilder:validation:Enum:=FailJobSet;Ignore;FailJob;RestartJob
+  Action FailurePolicyAction `json:"action"`
+  // The requirement on the job failure reasons. The requirement
+  // is satisfied if at least one reason matches the list.
+  // Each unique failure reason can only be associated with one FailurePolicyRule.
+  // An empty list applies the rule to any job failure reason.
+  // +kubebuilder:validation:UniqueItems:true
+  OnJobFailureReasons []string `json:"onJobFailureReasons"`
+  // TargetReplicatedJobs are the names of the replicated jobs the operator will apply to.
+  // A null or empty list will apply to all replicatedJobs.
+  // +optional
+  // +listType=atomic
+  // +kubebuilder:validation:UniqueItems
+  TargetReplicatedJobs []string `json:"targetReplicatedJobs,omitempty"`
+}
+
+type FailurePolicy struct {
+  // MaxRestarts defines the limit on the number of JobSet restarts.
+  // A restart is achieved by recreating all active child jobs.
+  MaxRestarts int32 `json:"maxRestarts,omitempty"`
+  // List of failure policy rules for this JobSet.
+  // Only one rule can match a unique (Job failure reason, ReplicatedJob)
+  // pair. If such a rule exists, it is executed. Otherwise, the default
+  // failure policy of attempting to restart the JobSet is applied.
+  // +optional
+  Rules []FailurePolicyRule `json:"rules,omitempty"`
+}
+
+```
+
+### Constraints
+
+- Each unique (Job failure reason, target Replicated Job) pair can only be associated with one FailurePolicyRule.
+This gives the user the flexibility to specify different failure and restart semantics for jobs failing for the same
+reason, depending on what Replicated Job they are part of. However, within a Replicated Job, there cannot be more than
+one way of handling a particular Job failure reason.
+
+- OnJobFailureReasons must be a valid Job failure reason as defined [here](https://github.com/kubernetes/kubernetes/blob/2d4100335e4c4ccc28f96fac78153f378212da4c/staging/src/k8s.io/api/batch/v1/types.go#L537-L554)
+in the batchv1 Job API. At the time of writing this KEP, these include:
+  - `PodFailurePolicy`
+  - `BackoffLimitExceeded`
+  - `DeadlineExceeded`
+  - `MaxFailedIndexesExceeded`
+  - `FailedIndexes`
+
+### Implmentation
+
+The core part of the implementation will be defining what specific mechanisms the JobSet controller uses to implement
+the behavior defined for each FailurePolicyAction type:
+
+1) `RestartJob`: To restart a single child job without restarting the entire JobSet, the controller will delete that 
+particular child job and allow the normal reconciliation process to recreate it.
+
+2) `FailJob`: To leave a particular child job in a failed state without restarting it or restarting the JobSet, the
```

Comment:
Question, say we have 2 Jobs failed. One matched the `FailJob` rule, the other does not match any rules, but we have `maxRestarts>0`, and it is not reached yet. Now, two cases, depending on the order of matching the Job failures:
1. [job1 (FailJob), job2(no matching rule)] - so after job1 we do nothing, but when we process job2, do we restart the JobSet, including Job1?
2. [job2 (no matching rule), job1 (FailJob)] - so we restart the JobSet, and then job1 (FailJob) is essentially ignored.

Is my understanding correct, is this expected behavior? I find it a little bit tricky in both cases actually.

Reply:
This is a good point - we need to think carefully about how to handle conflicting actions for different failed jobs which fail at the same time. 

I think broader failure policy rules should still apply to failed jobs which match narrower failure policies. 

In your example 1 above, the first job will be left failed, then the next job will trigger a recreate of the whole JobSet. That is the intended behavior.

In example 2 above, the first failed job will trigger a JobSet recreation, and the second failed job will also be restarted, regardless of the fact that it matched a narrower failure policy rule, which is the intended behavior.

The narrower failure policies are there to enable the user to avoid more expensive restart actions (e.g., recreating the whole JobSet) if there is **only** a failed job matching that rule (for example, enable the user not restart the whole JobSet if ONLY some auxiliary metrics collecting jobs failed).  

However, we still want the broader, more expensive restart policies to apply to all applicable jobs, including that failed job which matched some other failure policy. I'm not aware of a use case for a broad restart policy ignoring failed jobs which matched a narrower failure policy.

@ahg-g what do you think?

Reply:
Daniel, thanks for explaining. This makes sense. 

I'm still wondering if there would also be useful concept of a FailedJob (failed for good, not retried even on JobSet restarts)? 

Recently in the core k8s Job we added the FailIndex action which marks some indexes as no-longer retriable, even if the BackoffLimitPerIndex would still allow for the restarts. The analogy I see is that with the FailJob action here some users may expect that using "FailJob" action they fail it for good. Still, this could be another action like "NonRestartableFailJob", added later if needed.

Reply:
hmm, I was actually assuming that FailJob is similar to FailIndex, failed for good.